### PR TITLE
Read/Write PNG images using libpng

### PR DESCRIPTION
**Pull request by ricortiz**
This patch effectively removes devIL as a dependency and transfers the png read/write functionality to libpng.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -7,7 +7,7 @@ project(OpenTissue VERSION 0.994
 if(EXISTS ${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
   include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
   conan_basic_setup(TARGETS NO_OUTPUT_DIRS)
-  set(Boost_USE_STATIC_LIBS ON)
+  # set(Boost_USE_STATIC_LIBS ON)
```
Comment (erleben) [inline CMakeLists.txt]: Why keep out commented CMake code around?

File: OpenTissue/gpu/image/io/image_read.h [hunk 2]
```diff
@@ -1,91 +1,180 @@
...
-#define OPENTISSUE_GPU_IMAGE_IO_IMAGE_READ_H
 //
 // OpenTissue Template Library
 // - A generic toolbox for physics-based modeling and simulation.
 // Copyright (C) 2008 Department of Computer Science, University of Copenhagen.
 //
 // OTTL is licensed under zlib: http://opensource.org/licenses/zlib-license.php
 //
-#include <OpenTissue/configuration.h>
+#pragma once
```
Comment (erleben) [inline OpenTissue/gpu/image/io/image_read.h:8]: Does this really work on Mac/Linux?

As I recall this is non-standard?

The purpose of the ifndef and define blocks are to achieve the same effect but using standard only

File: OpenTissue/gpu/image/io/image_read.h [hunk 3]
```diff
@@ -1,91 +1,180 @@
...
-#include <OpenTissue/gpu/image/io/image_il_wrap.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <png.h>
 
 #include <iostream>
 #include <string>
+#include <vector>
+
+#include <OpenTissue/configuration.h>
```
Comment (erleben) [inline OpenTissue/gpu/image/io/image_read.h:15]: Usually, the OT convention is to have most local includes first and more distance includes later.

That means OT files first, 3rd party stuff (bundled or nonbundled) next, and finally system dependent stuff last, like standard header libraries

File: OpenTissue/gpu/image/io/image_read.h [hunk 4]
```diff
@@ -1,91 +1,180 @@
...
+namespace image {
+/**
+* Read Image.
+*
+* @param   filename
+* @param   image
+* @return               If succesful then the return value is true
+*                       otherwise it is false.
+*/
+bool read(std::string const & filename,
```
Comment (erleben) [inline OpenTissue/gpu/image/io/image_read.h:28]: OT have used the convention of writing commas a little differently on the next line

File: OpenTissue/gpu/image/io/image_read.h [hunk 5]
```diff
@@ -1,91 +1,180 @@
...
+  if (!end_info)
+  {
+      std::cerr << "error: png_create_info_struct returned 0" << std::endl;
+      png_destroy_read_struct(&png, &info, nullptr);
+      fclose(fp);
+      return false;
+  }
+
+
+  if (setjmp(png_jmpbuf(png))) {
```
Comment (erleben) [inline OpenTissue/gpu/image/io/image_read.h]: Inconsistent formatting of curly braces

File: OpenTissue/gpu/image/io/image_write.h [hunk 6]
```diff
@@ -1,64 +1,117 @@
...
 // - A generic toolbox for physics-based modeling and simulation.
 // Copyright (C) 2008 Department of Computer Science, University of Copenhagen.
 //
 // OTTL is licensed under zlib: http://opensource.org/licenses/zlib-license.php
 //
-#include <OpenTissue/configuration.h>
-
-#include <OpenTissue/gpu/image/image.h>
-#include <OpenTissue/gpu/image/io/image_il_wrap.h>
+#pragma once
```
Comment (erleben) [inline OpenTissue/gpu/image/io/image_write.h:8]: Is this standard? I am concerned about cross-compiler support on multiple plaltforms

File: OpenTissue/gpu/image/io/image_write.h [hunk 7]
```diff
@@ -1,64 +1,117 @@
...
 
+#include <png.h>
```
Comment (erleben) [inline OpenTissue/gpu/image/io/image_write.h]: stl headers usually come last?

File: OpenTissue/gpu/image/io/image_write.h [hunk 8]
```diff
@@ -1,64 +1,117 @@
...
+namespace image {
+/**
+* Write Image.
+*
+* @param   filename
+* @param   image
+* @return  If successful then the return value is true
+*          otherwise it is false.
+*/
+bool write(std::string const & filename,
```
Comment (erleben) [inline OpenTissue/gpu/image/io/image_write.h:28]: , formating is non-OT convention

File: unit_tests/io/src/unit_read_png.cpp [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+#include <boost/test/tools/floating_point_comparison.hpp>
+#include <boost/test/test_tools.hpp>
+#include <OpenTissue/utility/utility_pop_boost_filter.h>
+
+BOOST_AUTO_TEST_SUITE(opentissue_io);
+
+BOOST_AUTO_TEST_CASE(test_cases)
+{
+  {
+    const std::string filename = "/home/rortiz/Projects/OpenTissue/build/unit_tests/io/pax.png";
```
Comment (erleben) [inline unit_tests/io/src/unit_read_png.cpp]: incorrect const convention, OT uses the std:string const filename convention

**Review by erleben — APPROVED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (ricortiz) [reply CMakeLists.txt]: this was an oversight...

File: OpenTissue/gpu/image/io/image_read.h (hunk 2 above)
Comment (ricortiz) [reply OpenTissue/gpu/image/io/image_read.h:8]: Yes, it is supported by the compilers that mater (gcc, clang, msvc), and yes is non-standard. It has become second nature for me to use the pragma. Will revert.

File: OpenTissue/gpu/image/io/image_read.h (hunk 3 above)
Comment (ricortiz) [reply OpenTissue/gpu/image/io/image_read.h:15]: Got it, will revert.

File: OpenTissue/gpu/image/io/image_read.h (hunk 4 above)
Comment (ricortiz) [reply OpenTissue/gpu/image/io/image_read.h:28]: Got it, will revert.

File: OpenTissue/gpu/image/io/image_read.h (hunk 5 above)
Comment (ricortiz) [reply OpenTissue/gpu/image/io/image_read.h]: hmm, are you reviewing my the first commit on this PR? this is fixed on the second commit.

File: CMakeLists.txt (hunk 1 above)
Comment (ricortiz) [reply CMakeLists.txt]: actually, I fixed this on the second commit as well.

File: OpenTissue/gpu/image/io/image_write.h (hunk 6 above)
Comment (ricortiz) [reply OpenTissue/gpu/image/io/image_write.h:8]: It is actually non-standard but well supported by most compilers. While more and more libraries are starting to use this I understand your concern.

File: unit_tests/io/src/unit_read_png.cpp (hunk 9 above)
Comment (ricortiz) [reply unit_tests/io/src/unit_read_png.cpp]: old commit...

**Review by erleben — APPROVED** [review]
Nice job

**Pull request merged**